### ISSUE TITLE
mysql default instance

### ISSUE THREAD

**Issue by sdarwin** [chef-cookbooks/mysql#344]
It seems like most of the time, you only need the default "instance" of mysql, and when that's the case, it would be convenient to utilize the standard paths for everything.

i.e. /var/run/mysql , /etc/mysql , /etc/init/mysql.conf , etc 

The final results of a chef cookbook are not only for DevOps chefs.    They are for the developers and the end-users too.   Is it necessary to explain to everybody, they must connect with `mysql -S /var/run/mysql-default/mysqld.sock` instead of the normal way, and figure out all the new paths, because chef has decided to implement "instances" ,which are not commonly required (are they?).   things should be as simple as possible, when they can be.

**Comment by sdarwin** [comment]
There must have been a reason that the mysql cookbook 6.x was written in this way, which unfortunately breaks compatibility with a standard mysql installation, and the previous cookbook.

Perhaps certain puzzles or difficulties presented themselves, and that was the easiest way to solve them at the moment.  What if we can list all those problems and then figure them out.

Here is one such problem:  what if you have a mysql instance "foo"?  should the default package instance then be deleted, or continue to run as a second mysql instance? 

There are three cases to consider:

Case 1:  only the default instance is running.   
Case 2:  the default instance is defined, and a second instance "foo" is also defined.  
Case 3:  the default instance is not defined.  A second instance "foo" is defined, or possibly also a third instance "bar", etc, 

For Case 3, it's a mystery, because the default instance may or may not conflict with the "foo" instance.   So what should happen?   

Most likely, the solution for Case 3 ought to be an error.   Chef will stop running, and print this error message:

"Please specify what you would like to have happen with default operating system mysql instance, by defining it at least in a minimal way, as follows:
mysql_service 'default' do
  action [:delete]
end
_OR_
mysql_service 'default' do
  action [:create]
  ... and possibly more information, such as port, data_dir, etc would go here ...
end
"

Then, implement the 'default' instance with all the operating system defaults as with the mysql cookbook 5.0

What this solution is offers, is 90% of chef customers and chef developers will get a default, standard mysql installation and not have to do anything usual.  Case 3 won't apply to them.   They won't see that error message.

And then maybe only 10% of chef developers have to jump through a small hoop, as the expression goes.  But that trade-off makes sense, rather than the other way around where everyone is affected.

**Comment by mstyne** [comment]
@sdarwin I'm also very curious as to the rationale behind this implementation. When there's a section in the FAQ titled 'The mysql command line doesn't work' -- that seems like a red flag.

While I've been able make the necessary updates to use this version of the community cookbook, I was surprised that what appears to be a very non-standard way of deploying MySQL is defined as the default way.

Cookbook authors, don't read this as a complaint -- it's really just a request for clarification on why this implementation was chosen. Your hard work and maintenance of this project is appreciated.

**Comment by eric-tucker** [comment]
I agree with the sentiment expressed here.  While I am aware of this intricacy and the (new) need to specify the socket on the command line, the developers and testers will not be aware of this.  I know DevOps is about bringing everyone up to speed and communication, but it won't always be possible.

I too would appreciate if a default installation were to work exactly as the default package install works so that no additional command line options were needed:

````
mysql -u root -p
````

This should work out of the box but it does if not using this cookbook.

While this isn't the end of the world, it would be nice if this behaviour could be maintained.

**Comment by dazwin** [comment]
+1 on the request and sentiments expressed here

Multiple MySQL instances is an edge case. I'm afraid offering the -S option is a poor 'get out' as it fails to consider that other applications and libraries must now be reconfigured where default behavior was just fine before. In some of these cases, connecting via unix socket has not be exposed (the 'localhost' switcheroo is relied upon) and we must now deal with a real network connection.

Please supplement the good work of the .deb/.rpm packages (rather than replace it) as the default option, and support multiple MySQL instances only if explicitly needed.

Thanks!

**Comment by someara** [comment]
Hi.

There's actually a quite a bit to cover here, and I'll do my best to
get to it all...

TLDR - "Just connect over the TCP socket."

5.x implemented a "singleton" resource, which did not behave like a
normal Chef resource. Normally, when you declare multiple instances of
a thing in a recipe, you get multiple instances of the thing.

For example, the user resource:

```
user 'alice'
user 'bob'
user 'carol'
```

Or, the directory resource:

```
directory '/foo'
directory '/bar'
directory '/biz'
```

In 5.x, a recipe with multiple mysql_service resources would only
yield the single service, creating confusion, and departing from the
normal behavior that a user would expect from a recipe.

```
mysql_service 'foo'
mysql_service 'bar'
mysql_service 'biz'
```
In 6.x, this works as expected.

Running multiple instances of a service on a machine isn't as big of
an edge case as you might think. It's actually fairly common. Cramming
multiple instances onto a machine with lots of cores lets you
consolidate resources for dev and testing environments. Sharding into
multiple instances lets you overcome the single threaded replication
limitation in older versions. (pre 5.6)

During development, I initially attempted to make
mysql_service[default] special, in an attempt to use native paths.

This method: http://bit.ly/1dEPc59
... originally had the line:
`return 'mysql' if new_resource.instance == 'default'`

I was of the mind that adhering to the platform native idioms was the
best thing to do. It would certainly fix the mysql client socket path
issue. I quickly discovered that this was not the case.

Popular Linux distributions' native idioms and "helpful" features are
designed aide a manual administration model. In many cases, they're
actively hostile to automated CM.

The Oracle distributed RPMs drop of a directory structure (var, etc),
then the post scripts initialize the database and set a random
password for the root user. The password is stashed in a file on disk.
During the "first boot", you're supposed to start the service, read
the password from the file, connect to database, and  change it some a
known value. If you want to use a different datadir, you're supposed
to reconfigure the service and move the datadir contents over.

When configuring a NEW machine that utilizes point to a pre-existing
database, (mounting a network share or block device into data_dir),
there is no way to know if we should perform "first run" procedures.

The Debian packaging is even worse. Like the Enterprise Linux
packages, Debian postinst initializes the database, and sets a
password. It goes the extra mile and starts the service, giving no
chance to reconfigure datadir.

Don't even get me started on debian-sys-maint.

Package upgrades often contain unpredictable post installation
scripts, override changes to init scripts, and place extra files in
configuration directories.

It turns out what we want is full, manual control over every step of
the configuration process. Of all the platforms this cookbook
supports, OmniOS turned out to be the easiest to deal with. The
"platform native idioms" embodied in the packaging can be summed up as
"dump a pre-compiled binary to disk, and then let the user sort it
out." There was no package post scripts, no service manager
registration, no opinionated pathing, nothing. Beautiful.

One last thought on platform idioms... there is no reason for the "one
instance of a service per machine" outside of it being complicated to
enable using shell-script driven service management schemes.

This cookbook ships a resource that behaves the same using 5 MySQL
versions times 12 OS/version combinations. It turns out, that's hard.
I can't remember the exact details, but somewhere along the line,
there was an edge case where MySQL would use /etc/my.cnf, if present,
ignoring the --defaults-file entirely.

So please. Pretty please, with sugar on top. Just the TCP socket.

**Comment by someara** [comment]
PS, you can use the "socket" property in the mysql_service resource to specify /var/lib/mysql/mysql.sock if you want, making the mysql client work without extra args.

**Comment by ConfusedVorlon** [comment]
another workaround here is to symlink the default socket to the specific one

```
directory '/var/run/mysqld' do
  owner 'mysql'
  group 'mysql'
  mode '0755'
  action :create
end

link '/var/run/mysqld/mysqld.sock' do
  to '/var/run/mysql-foo/mysqld.sock'
end
```
